File: android/app/src/main/java/com/modi/connect/audio/AudioConfig.kt
/*
 * MoDi Connect - Cross-device interconnection protocol
 * Copyright (C) 2026 Silvite
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <https://www.gnu.org/licenses/>.
 */
package com.modi.connect.audio

/**
 * AudioConfig — Android 端音频参数集中配置
 *
 * ## 职责
 * 统一管理所有音频相关常量，提供默认参数。
 * 各模块通过构造函数接收此配置，消除分散的 hardcode。
 *
 * ## 默认值说明
 * - 采样率 48kHz / 16bit / 单声道 — 通用 AC 级语音/音乐传输
 * - 帧长 20ms（960 采样点）— Opus 推荐帧长，延迟与压缩率平衡
 * - Opus 码率 128kbps — 局域网带宽充足，音质优先
 * - 复杂度 10（最高）— 本机编码不计较 CPU
 * - FEC 开启 + 15% 预期丢包率 — 抗 WiFi 不稳定
 * - CVBR — 约束 VBR，为 FEC 留余量
 *
 * ## 缓冲说明
 * - MicrophoneCapturer: MinBufferMultiplier=4（~80ms），抗 AudioRecord 读取抖动
 * - SystemAudioCapturer: MinBufferMultiplier=32（~640ms），抗 MediaProjection 初始丢帧
 */
data class AudioConfig(
    /** 采样率（Hz），默认 48000 */
    val sampleRate: Int = 48000,
    /** 声道数，默认 1（单声道） */
    val channels: Int = 1,
    /** 帧长（毫秒），默认 20ms */
    val frameMs: Int = 20,
    /** Opus 编码码率（bps），默认 128000 */
    val bitrate: Int = 128000,
    /** Opus 编码复杂度（0-10），默认 10 */
    val complexity: Int = 10,
    /** 是否启用带内 FEC，默认 true */
    val useFec: Boolean = true,
    /** 预期丢包率百分比，默认 15 */
    val packetLossPercent: Int = 15,
    /** 是否使用约束 VBR，默认 true */
    val useConstrainedVbr: Boolean = true,
    /** Opus 信号类型，默认 OPUS_SIGNAL_MUSIC */
    val signalType: Int = 0,  // 0 = OPUS_SIGNAL_MUSIC
    /** 麦克风采集缓冲倍数，默认 4 */
    val micBufferMultiplier: Int = 4,
    /** 系统音频采集缓冲倍数，默认 32 */
    val sysAudioBufferMultiplier: Int = 32,
) {
    // ── 派生常量 ──

    /** 每帧采样点数 */
    val frameSize: Int get() = sampleRate * frameMs / 1000
    /** 每帧字节数（PCM16LE） */
    val frameBytes: Int get() = frameSize * 2  // 16bit = 2 字节

    companion object {
        /** 默认配置 */
        val DEFAULT = AudioConfig()
    }
}

File: windows/MoDi.Desktop/AudioConfig.cs
/*
 * MoDi Connect - Cross-device interconnection protocol
 * Copyright (C) 2026 Silvite
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <https://www.gnu.org/licenses/>.
 */
namespace MoDi.Desktop;

/// <summary>
/// AudioConfig — Windows 端音频参数集中配置
///
/// ## 职责
/// 统一管理所有音频相关常量，消除散落在各文件中的 hardcode。
/// AudioRouter 通过构造函数接收此配置。
///
/// ## 延迟参数说明（当前值，经实测稳定）
/// - BufferMs = 100 — 抖动缓冲，从 300 逐步降至 100，再降需监控 UNDERRUN
/// - WaveOutLatency = 100 — WaveOutEvent DesiredLatency，50 卡顿回退至此
/// - WasapiLatency = 50 — WasapiOut（CABLE Input），50 稳定的下限
///
/// ## 音频参数
/// - 采样率 48kHz / 16bit / 单声道 / 20ms 帧（960 采样点 / 1920 字节）
/// - Opus 解码对应 128kbps 编码流
/// </summary>
public sealed class AudioConfig
{
    /// <summary>采样率（Hz）</summary>
    public int SampleRate { get; set; } = 48000;

    /// <summary>声道数（单声道）</summary>
    public int Channels { get; set; } = 1;

    /// <summary>每帧时长（ms）</summary>
    public int FrameMs { get; set; } = 20;

    /// <summary>UDP 音频端口</summary>
    public int AudioPort { get; set; } = TransportIdentity.AudioPort;

    // ── 缓冲/延迟参数 ──

    /// <summary>抖动缓冲时长（ms）</summary>
    public int BufferMs { get; set; } = 100;

    /// <summary>WaveOutEvent DesiredLatency（ms）</summary>
    public int WaveOutLatency { get; set; } = 100;

    /// <summary>WasapiOut CABLE Input DesiredLatency（ms）</summary>
    public int WasapiLatency { get; set; } = 50;

    /// <summary>音频看门狗超时（ms），超过此时间无音频包认为连接中断</summary>
    public int AudioTimeoutMs { get; set; } = 3000;

    // ── 派生常量 ──

    /// <summary>每帧采样点数</summary>
    public int FrameSize => SampleRate * FrameMs / 1000;

    /// <summary>每帧字节数（PCM16LE）</summary>
    public int FrameBytes => FrameSize * sizeof(short);

    /// <summary>默认配置</summary>
    public static readonly AudioConfig Default = new();
}
